gtdynamics/dynamics/Chain.h

```diff
@@ -0,0 +1,84 @@
+/* ----------------------------------------------------------------------------
+ * GTDynamics Copyright 2021, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * See LICENSE for the license information
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file Chain.h
+ * @brief Create a serial kinematic chain.
+ * @author Dan Barladeanu, Frank Dellaert.
+ */
+
+
+#pragma once
+
+//#include <gtsam/linear/NoiseModel.h>
+//#include <gtsam/nonlinear/NonlinearFactorGraph.h>
+//#include <gtsam/nonlinear/Values.h>
+#include <boost/optional.hpp>
+#include <gtsam/geometry/Pose3.h>
+#include <gtsam/base/Matrix.h>
+#include <gtsam/base/Vector.h>
+
+using gtsam::Pose3;
+using gtsam::Matrix;
+using gtsam::Vector;
+
+namespace gtdynamics {
+
+/**
+ * Chain is a class used to create a serial kinematic chain, which 
+ * helps in creating a LeanDynamicsGraph
+ */
+class Chain {
+ private:
+  Pose3 sMb_; //rest pose of "body" with respect to "spatial" frame.
+  Matrix axes_; //screw axes of all joints in the chain expressed in body frame.
+
+ public:
+  /// Constructor
+  Chain(Pose3 &sMb, Matrix &axes) : sMb_(sMb), axes_(axes) {}
+  
+  /**
+   * This function performs the monoid operation on pose + jacobian pairs.
+   * The monoid operation returns the composed chain pose and jacobian.
+   * Detailed explanation in readme (chain.md)
+   * @param aTb ............. chain 1 pose
+   * @param bAj ............. chain 1 screw axes
+   * @param bTc ............. chain 2 pose
+   * @param cAk ............. chain 2 screw axes
+   * @param aTc ............. composed chain pose
+   * @param cAjk ............ composed chain jacobian
+   */
+  static void monoidOperation(const Pose3 &aTb, const Matrix &bAj, const Pose3 &bTc,
+                       const Matrix &cAk, Pose3 &aTc, Matrix &cAjk);
```

Comment:
output parameters should be pointers, see Google style guide, *but* I'd rather have something functional like
```
static Chain monoidOperation(const Chain& a, const Chain& b);
```

Reply:
2nd this. I am not a fan of output parameters. It is error prone and not functional.

Reply:
changed this to overloaded star operator with chains as inputs
done